### PR TITLE
PushQueue perf optimization

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v2/pushqueue.go [hunk 1]
```diff
@@ -63,17 +63,23 @@ func (p *PushQueue) Enqueue(proxy *XdsConnection, pushInfo *PushEvent) {
...
+				edsUpdates[endpoint] = struct{}{}
+			}
+			event.edsUpdatedServices = edsUpdates
+		} else {
+			event.edsUpdatedServices = nil
 		}
-		event.edsUpdatedServices = edsUpdates
 	}
-	p.cond.Signal()
+	if len(p.connections) > 0 {
```
Comment (howardjohn) [inline pilot/pkg/proxy/envoy/v2/pushqueue.go]: I think this is always true

Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/pushqueue.go]: Ah, i must be drunk.

**Comment by nmittler** [comment]
@hzxuzhonghu can you run a benchmark to show the performance difference of the code that uses this?

**Comment by hzxuzhonghu** [comment]
@nmittler Yes, i can do that. It depends on the number of edsUpdatedServices.

**Comment by hzxuzhonghu** [comment]
```

func BenchmarkEnqueue(b *testing.B) {
	con := &XdsConnection{ConID: "proxy"}
	edsUpdatedServices1 := map[string]struct{}{}
	edsUpdatedServices2 := map[string]struct{}{}

	for i := 0; i < 20; i++ {
		if i < 10 {
			edsUpdatedServices1[fmt.Sprintf("svc-%d", i)] = struct{}{}
		} else {
			edsUpdatedServices2[fmt.Sprintf("svc-%d", i)] = struct{}{}

		}
	}

	for i := 0; i < b.N; i++ {
		p := NewPushQueue()
		p.Enqueue(con, &PushEvent{full:true, edsUpdatedServices: edsUpdatedServices1})
		p.Enqueue(con, &PushEvent{full:true, edsUpdatedServices: edsUpdatedServices2})

	}
}
```


Without this pr: 
```
goos: linux
goarch: amd64
pkg: istio.io/istio/pilot/pkg/proxy/envoy/v2
BenchmarkEnqueue-32    	  200000	      6477 ns/op
PASS
ok  	istio.io/istio/pilot/pkg/proxy/envoy/v2	3.476s
```

With this:

```
goos: linux
goarch: amd64
pkg: istio.io/istio/pilot/pkg/proxy/envoy/v2
BenchmarkEnqueue-32    	 1000000	      1233 ns/op
PASS
ok  	istio.io/istio/pilot/pkg/proxy/envoy/v2	3.361s
```

**Comment by nmittler** [comment]
@hzxuzhonghu 

Thanks for posting the results ... could you also include a version of your benchmark so we can track the performance going forward?

**Comment by hzxuzhonghu** [comment]
>could you also include a version of your benchmark so we can track the performance going forward?

Do you mean the golang version? 
```
root@beast-02:~# go version
go version go1.12.5 linux/amd64
```

**Pull request merged**